code/krepel/math/vector3.d

```diff
@@ -0,0 +1,479 @@
+module krepel.math.vector3;
+
+import krepel.math.math;
+import krepel.algorithm.comparison;
+import std.conv;
+
+@nogc:
+@safe:
+
+/// Calculates the Dot Product of the two vectors
+/// Input vectors will not be modified
+float Dot(Vector3 lhs, Vector3 rhs)
+{
+  float result = lhs | rhs;
+  return result;
+}
+
+/// Multiplies two vectors per component returning a new vector containing
+/// (lhs.X * rhs.X, lhs.Y * rhs.Y, lhs.Z * rhs.Z)
+/// Input vectors will not be modified
+Vector3 Mul(Vector3 lhs, Vector3 rhs)
+{
+  return lhs * rhs;
+}
+
+/// Calculates the cross Product (lhs x rhs) and returns the result
+/// Input vectors will not be modified
+Vector3 Cross(Vector3 lhs, Vector3 rhs)
+{
+  return lhs ^ rhs;
+}
+
+/// Calculates the squared length  of the given vector, which is the same as the dot product with the same vector
+/// Input vector will not be modified
+float LengthSquared(Vector3 vec)
+{
+  return vec | vec;
+}
+
+/// Calculates the squared length of the X and Y components, ignoring the Z component
+/// Input vector will not be modified
+float LengthSquared2D(Vector3 vec)
```

Comment:
Maybe the `-2D` version become obsolete with the swizzling operators? `v3.LengthSquared2D()` vs. `v3.XY.LengthSquared()`.